prometheus/quantile_tracker.go

```diff
@@ -0,0 +1,233 @@
+		t.uh = nest - t.est + (1-t.rho)*t.uh + t.rho*v
+		t.ul = nest - t.est + t.ul
+	} else {
+		t.uh = nest - t.est + t.uh
+		t.ul = nest - t.est + (1-t.rho)*t.ul + t.rho*v
+	}
+	t.est = nest
+}
+
+// MQTracker is an implementation of:
+// Joint Tracking of Multiple Quantiles Through Conditional Quantiles
+// Hammer, Yazidi, and Rue
+// https://arxiv.org/pdf/1902.05428.pdf
+// This version allows substituting different internal trackers.
+type MQTracker struct {
+	est       []float64 // Current estimates
+	tr        []Tracker // Individual trackers
+	K         int       // length(quantiles)
+	c         int       // Index of position quantile
+	quantiles []float64 // Quantiles to track
+	lam       float64   // Position convergence rate parameter
+	gam       float64   // Shape convergence rate parameter
+	rho       float64   // Conditional expectation convergence rate parameter
+}
+
+// Estimate returs a slice containing the current estimates.
+// Note: for threading purposes, this is a copy, thus can be modified.
+// However, the function itself is not threadsafe, so should be called under
+// a lock.
+func (t *MQTracker) Estimate() []float64 {
+	est := append(t.est[:0:0], t.est...)
+	return est
+}
+
+// Observe inserts new data.
+// Note: not threadsafe, should be called under a lock.
+func (t *MQTracker) Observe(v float64) {
+	t.tr[t.c].Observe(v)
+	t.est[t.c] = t.tr[t.c].Estimate()
+	for k := t.c - 1; k >= 0; k-- {
+		nest := t.est[k+1]
+		if v < nest {
+			y := v - nest
+			t.tr[k].Observe(y)
+		}
+		t.est[k] = t.tr[k].Estimate() + nest
+	}
+	for k := t.c + 1; k < t.K; k++ {
+		nest := t.est[k-1]
+		if v > nest {
+			y := v - nest
+			t.tr[k].Observe(y)
+		}
+		t.est[k] = t.tr[k].Estimate() + nest
+	}
+}
+
+// Generic constructor, not exported.
+func newMTracker(maker TrackerMaker, q []float64, lam, gam, rho float64) *MQTracker {
+	K := len(q)
+	// Find the index of the quantile closest to the median.
+	// This will be the one that tracks the position of the distribution.
+	// Override after construction if necessary (can be a good idea if you
+	// know the distribution has a mass at one end of the support, e.g.
+	// use 0 if tracking latency).
+	var c int
+	var min = 2.0 // Deliberately crazy value.
+	for i, v := range q {
+		v2 := math.Abs(v - 0.5)
+		if v2 < min {
+			min = v2
+			c = i
+		}
+	}
+	r := MQTracker{
+		make([]float64, len(q)),
+		make([]Tracker, len(q)),
+		K, c,
+		make([]float64, len(q)),
+		lam, gam, rho}
+	for i := 0; i < K; i++ {
+		r.est[i] = 0
+	}
+	r.quantiles[c] = q[c]
+	for k := c - 1; k >= 0; k-- {
+		r.quantiles[k] = q[k] / q[k+1]
+	}
+	for k := c + 1; k < K; k++ {
+		r.quantiles[k] = (q[k] - q[k-1]) / (1 - q[k-1])
+	}
+	for i := 0; i < K; i++ {
+		l := gam
+		if i == c {
+			l = lam
+		}
+		r.tr[i] = maker(r.quantiles[i], l, rho, r.est[i])
+	}
+	return &r
+}
+
+// NewMQEMATracker constructs a MQTracker instance using QEMA inner trackers.
+func NewMQEMATracker(q []float64, lam, gam, rho float64) *MQTracker {
```

Comment:
This one is only used in the tests and benchmarks. Do you plan to use this one as an option in the actual Summary?